pulsar/producer_partition.go

```diff
@@ -784,57 +784,57 @@ func (p *partitionProducer) ReceivedSendReceipt(response *pb.CommandSendReceipt)
 		return
 	}
 
-	if pi.sequenceID != response.GetSequenceId() {
+	if pi.sequenceID < response.GetSequenceId() {
```

Comment:
this change will be reduce the number of reconnections only. this is very important.

Reply:
Now there are 2000 pod to consume one topic, mean there are 2000 consumers for one partitions, if this code is not bean changed , many times consumers need to be restarted.
Another words , this handle way is different with java sdk.

Reply:
@baomingyu The java logic as follows:

```
if (sequenceId > op.sequenceId) {
                log.warn("[{}] [{}] Got ack for msg. expecting: {} - {} - got: {} - {} - queue-size: {}", topic, producerName,
                        op.sequenceId, op.highestSequenceId, sequenceId, highestSequenceId, pendingMessages.size());
                // Force connection closing so that messages can be re-transmitted in a new connection
                cnx.channel().close();
                return;
            } else if (sequenceId < op.sequenceId) {
                // Ignoring the ack since it's referring to a message that has already timed out.
                if (log.isDebugEnabled()) {
                    log.debug("[{}] [{}] Got ack for timed out msg. expecting: {} - {} - got: {} - {}", topic, producerName,
                            op.sequenceId, op.highestSequenceId, sequenceId, highestSequenceId);
                }
                return;
            } else {
                // Add check `sequenceId >= highestSequenceId` for backward compatibility.
                if (sequenceId >= highestSequenceId || highestSequenceId == op.highestSequenceId) {
                    // Message was persisted correctly
                    if (log.isDebugEnabled()) {
                        log.debug("[{}] [{}] Received ack for msg {} ", topic, producerName, sequenceId);
                    }
                    pendingMessages.remove();
                    releaseSemaphoreForSendOp(op);
                } else {
                    log.warn("[{}] [{}] Got ack for batch msg error. expecting: {} - {} - got: {} - {} - queue-size: {}", topic, producerName,
                            op.sequenceId, op.highestSequenceId, sequenceId, highestSequenceId, pendingMessages.size());
                    // Force connection closing so that messages can be re-transmitted in a new connection
                    cnx.channel().close();
                    return;
                }
            }
```

In the previous processing, when the sequence ID is greater than the sequenceID of the response, we also close the current connection, expecting it to restore the logic here through internal retry.

The current implementation of pr ignores the case of `pi.sequenceID <response.GetSequenceId()`. When `pi.sequenceID <response.GetSequenceId()`, we should return directly here

Reply:
yes ，When pi.sequenceID <response.GetSequenceId() need to return.
Now code change is equal return directly way here.

Reply:
@cckellogg It seems that in some scenarios, the condition of `pi.sequenceID <response.GetSequenceId()` will be established, causing the connection to be closed. And @baomingyu they have more consumer objects under one connection, so when the reconnection happens, the impact on their business itself is magnified?

Reply:
pi.sequenceID < response.GetSequenceId() is error case，the channel must be closed when it has happed. And
pi.sequenceID > response.GetSequenceId() case should not do close way.

Reply:
So are you suggesting we follow the logic of the java client and handle the 3 cases this way?

1 - pi.sequenceID > response.GetSequenceId() => return (could there ever be a pending item in the queue that gets left around?)

2 - pi.sequenceID < response.GetSequenceId() => error close connection

3 - pi.sequenceID == response.GetSequenceId() => process